DROOLS-4689/DROOLS-4690/DROOLS-4691/DROOLS-4692: [DMN Designer] Code Completion - Enable code completion for Literal Expressions

karreiro opened the pull request:
See:
- [DROOLS-4689](https://issues.jboss.org/browse/DROOLS-4689): [DMN Designer] Code Completion - Setup Monaco (loader)
- [DROOLS-4690](https://issues.jboss.org/browse/DROOLS-4690): [DMN Designer] Code Completion - Add Monaco support for 'TextAreaSingletonDOMElementFactory'
- [DROOLS-4691](https://issues.jboss.org/browse/DROOLS-4691): [DMN Designer] Code Completion - Enable code completion for only Literal Expressions
- [DROOLS-4692](https://issues.jboss.org/browse/DROOLS-4692): [DMN Designer] Code Completion - UI adjusts for Monaco component into DMN Boxed Expression component

---

![2019-11-07 14 31 45](https://user-images.githubusercontent.com/1079279/68412704-74bea100-016b-11ea-80b8-c421d9c337c7.gif)

---

Part of an ensemble:
- https://github.com/kiegroup/appformer/pull/836
- https://github.com/kiegroup/kie-wb-common/pull/3000

karreiro commented:
Jenkins please retest this.

manstis commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/widgets/codecompletion/MonacoPropertiesFactory.java:
```diff
@@ -0,0 +1,282 @@
+                                    final String insertText) {
+
+        final JSONObject suggestion = makeJSONObject();
+        final int CompletionItemKindFunction = 1;
```
Camel case please.

karreiro replied:
Fixed. Thanks!

manstis commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/widgets/codecompletion/MonacoPropertiesFactory.java:
```diff
@@ -0,0 +1,282 @@
+
+        final JSONObject suggestion = makeJSONObject();
+        final int CompletionItemKindFunction = 1;
+        final int CompletionItemInsertTextRuleInsertAsSnippet = 4;
```
Camel case please.

karreiro replied:
Fixed. Thanks!

manstis commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/widgets/grid/columns/factory/AutocompleteTextareaDOMElementFactory.java:
```diff
@@ -0,0 +1,105 @@
+import org.uberfire.ext.wires.core.grids.client.widget.grid.keyboard.KeyDownHandlerCommon;
+import org.uberfire.ext.wires.core.grids.client.widget.layer.GridLayer;
+
+public class AutocompleteTextareaDOMElementFactory extends BaseSingletonDOMElementFactory<String, MonacoEditorWidget, MonacoEditorDOMElement> {
```
Sorry, could you make `TextArea` Camel case please.

karreiro replied:
Fixed. Thanks!

manstis commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/widgets/codecompletion/MonacoPropertiesFactoryTest.java:
```diff
@@ -0,0 +1,512 @@
+    }
+
+    @Test
+    public void testGetSuggestions() {
```
I'm sure this method could benefit from use of arrays... but it's fine as it is really.. just a comment.

karreiro replied:
Fixed. Thanks!

manstis commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/widgets/grid/columns/factory/dom/MonacoEditorDOMElementTest.java:
```diff
@@ -0,0 +1,215 @@
+        }
+
+        @Override
+        protected SimplePanel getContainer() {
```
Is this override needed?

karreiro replied:
Yes, otherwise we cannot mock `getContainer` on `testSetupContainerComponent`.

manstis commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/widgets/grid/BaseExpressionGrid.java:
```diff
@@ -252,6 +253,16 @@ public TextAreaSingletonDOMElementFactory getBodyTextAreaFactory() {
                                                       newCellHasValueCommand());
     }
 
+    public AutocompleteTextareaDOMElementFactory getAutocompleteTextareaFactory() {
```
Could you please Camel case `TextArea`.

karreiro replied:
Fixed. Thanks!

manstis commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/expressions/ExpressionContainerGridTest.java:
```diff
@@ -195,6 +196,9 @@
     private TextAreaSingletonDOMElementFactory textAreaSingletonDOMElementFactory;
 
+    @Mock
+    private AutocompleteTextareaDOMElementFactory autocompleteTextareaDOMElementFactory;
```
Sorry.. more camel case woes... I just won't be able to sleep!

karreiro replied:
Fixed. Thanks!

manstis commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/widgets/grid/columns/factory/AutocompleteTextareaDOMElementFactoryTest.java:
```diff
@@ -0,0 +1,152 @@
+import static org.mockito.Mockito.when;
+
+@RunWith(LienzoMockitoTestRunner.class)
+public class AutocompleteTextareaDOMElementFactoryTest extends BaseSingletonDOMElementFactoryTest<AutocompleteTextareaDOMElementFactory, MonacoEditorDOMElement> {
```
Camel case `TextArea`.

karreiro replied:
Fixed. Thanks!

manstis commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/widgets/grid/columns/factory/AutocompleteTextareaDOMElementFactoryTest.java:
```diff
@@ -0,0 +1,152 @@
+        verify(domElement).setupElements();
+    }
+
+    class AutocompleteTextareaDOMElementFactoryFake extends AutocompleteTextareaDOMElementFactory {
```
`TextArea`

karreiro replied:
Fixed.

manstis commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/widgets/codecompletion/MonacoPropertiesFactory.java:
```diff
@@ -0,0 +1,282 @@
+        return tokenizer;
+    }
+
+    public JSONArray getRoot() {
```
It might be nice to cite where these _root_ definitions come from in a JavaDoc?

karreiro replied:
I've added the Java Doc, Mike. Please let me know if it's clear now :-)

manstis commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/widgets/codecompletion/MonacoPropertiesFactory.java:
```diff
@@ -0,0 +1,282 @@
+        return root;
+    }
+
+    public JSONArray getSuggestions() {
```
It might be nice to cite where these suggestions come from in a JavaDoc?

e.g. are they all FEEL functions? If so where's the definitive list? In the OMG DMN1.2 specification?

karreiro replied:
No. This PR includes just part of the list. The complete list will be added by [DROOLS-4733](https://issues.jboss.org/browse/DROOLS-4733) :-)

karreiro commented:
Thank-you, @manstis :-)

karreiro commented:
@manstis. Thank you for your first review, but I've slightly change this PR with two commits since you approved it.

I got the final list of functions approved by Matteo this morning, so I've updated the list into one commit. And, in another commit, I've fixed the issue related to duplicated suggestions by initialising FEEL language once :-)

jomarko commented:
jenkins execute full downstream build

karreiro commented:
(rebasing the branch... 👀)

karreiro commented:
Jenkins, please execute full downstream build.

karreiro commented:
@jomarko I'm not sure if the full downstream is stable. So, I'm leaving the [.war](https://www.dropbox.com/s/tgpqz24f377b64k/kie-wb-common-dmn-webapp-standalone.war?dl=0) here :-)

karreiro commented:
Jenkins execute compile downstream build.

karreiro commented:
Jenkins, please execute full downstream build.

jomarko commented:
### Manual review note 1
Once completions are shown, then pressing **Escape** hides them however cancel also edit mode of the cell, I am not sure what customers will think about this.

karreiro commented:
@jomarko Personally, this seems like something expected (when I think as a user). But, I've filed a JIRA (https://issues.jboss.org/browse/DROOLS-4752) to iterate over this separately.

jomarko commented:
### Manual review note 2
Still under investigation, but posting here to give more time for fix. If I confirm, it is blocker from merge.

If you build **kie-wb-distributions** there is an error right after start of **business-central**

![Screenshot from 2019-11-13 16-41-17](https://user-images.githubusercontent.com/8044780/68780094-029bff80-0636-11ea-972b-4002854d9db2.png)

jomarko commented:
I do not continue with code review until the Note 2 is fixed.

karreiro commented:
Jenkins, please execute full downstream build.

karreiro commented:
@jomarko Thank you for reporting. The issue on **kie-wb-distributions** is fixed in a separated commit. Here's the [.war for kie-wb-distributions (business-central-webapp)](https://www.dropbox.com/s/5ajitpafknva4as/business-central-webapp.war?dl=0) and here's the [.war for kie-wb-common-dmn-webapp-standalone](https://www.dropbox.com/s/tgpqz24f377b64k/kie-wb-common-dmn-webapp-standalone.war?dl=0).

jomarko commented:
### Manual review note 3
The [video](https://issues.jboss.org/secure/attachment/12457469/auto-vs-explicit.webm) demonstrates different behavior when completions are shown automatically (first 2 attempts) and explicitly (last 2 attempts). Is that expected?

jomarko commented:
### Manual review note 4
Completions are not shown for decision tables. Acceptance criteria claims works for literal expressions. The cell in picture bellow is literal expression. Am I missing something?
![Screenshot from 2019-11-14 08-35-13](https://user-images.githubusercontent.com/8044780/68836121-e346b600-06b9-11ea-8731-28e4811acfb0.png)

jomarko commented:
### Manual review note 5
Same as note 4, but Relation

jomarko commented:
### Manaul review note 6
@manstis https://issues.jboss.org/browse/KOGITO-598

jomarko commented:
### Manaul review note 7
Each completions is shown [twice](https://issues.jboss.org/secure/attachment/12457471/completion-twice.webm)

jomarko commented:
### Manaul review note 8
Completions for last context entry are not visible.
[video](https://issues.jboss.org/secure/attachment/12457472/last-entry-completion.webm)

jomarko commented:
### Manual review note 9
Tab was previously used to finish edit mode. Now it finish the edit mode, however also append whitespace into the cell before.

jomarko commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/widgets/grid/columns/factory/dom/MonacoEditorDOMElement.java:
```diff
@@ -0,0 +1,179 @@
+        style.setPaddingLeft(5, PX);
+        style.setPaddingRight(5, PX);
+        style.setPaddingTop(5, PX);
+        style.setPaddingBottom(5, PX);
```
Isn't this just `style.setPadding(5, PX);` ?

karreiro replied:
Fixed. Thanks.

jomarko commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/widgets/grid/columns/factory/dom/MonacoEditorDOMElement.java:
```diff
@@ -0,0 +1,179 @@
+        final int rowIndex = context.getRowIndex();
+        final int columnIndex = context.getColumnIndex();
+
+        if (value == null || value.trim().isEmpty()) {
```
This is `if (StringUtils.isEmpty(value)) {`

karreiro replied:
Fixed. Thanks.

jomarko commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/widgets/grid/columns/factory/dom/MonacoEditorWidget.java:
```diff
@@ -0,0 +1,66 @@
+        getCodeEditor().ifPresent(c -> c.setValue(value));
+    }
+
+    public String getValue() {
```
@Override please

karreiro replied:
Fixed. Thanks.

jomarko commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/widgets/grid/columns/factory/dom/MonacoEditorWidget.java:
```diff
@@ -0,0 +1,66 @@
+        return Optional.ofNullable(codeEditor);
+    }
+
+    public void setTabIndex(final int index) {
```
@Override please

karreiro replied:
Fixed. Thanks.

jomarko commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/widgets/codecompletion/MonacoPropertiesFactoryTest.java:
```diff
@@ -0,0 +1,533 @@
+        verify(options).put("useTabStops", useTabStops);
+        verify(options).put("contextmenu", contextmenu);
+        verify(options).put("folding", folding);
+        verify(miniMap).put("enabled", enabled);
```
I think we shouldn't mix assertions related to `options` and related to `minimap` and `scrolbar` due to readability

karreiro replied:
Thank you, @jomarko. But I think the assertions are clear as they are (considering the context of the properties). Thanks again for noticing it.

jomarko commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/widgets/codecompletion/MonacoPropertiesFactoryTest.java:
```diff
@@ -0,0 +1,533 @@
+        verify(factory).push(expectedRules, rule5);
+
+        assertEquals(expectedRules, actualRules);
+    }
```
please update the variable names to reflect what they are used for, also in the source class

karreiro replied:
Thank you, @jomarko. But I prefer to keep it as is. In the source class the name of the variables are: `rule1`, `rule2`, and so on. Here's I creating a context by having `foreground1`, `token2`, and so on. Thus, I'm keeping it as is. Thanks again for noticing it.

jomarko commented:
Done with review, the ensemble **can not be merged**, because it breaks the existing selenium [tests](https://github.com/jboss-integration/bxms-qe-tests/blob/master/test-kie-wb/src/test/java/org/jboss/qa/brms/dmn/DmnTest.java).

jomarko commented:
### Manual review note 10
There is way context menu and completions are [shown simultaneously](https://issues.jboss.org/secure/attachment/12457479/context-completions.webm).

karreiro commented:
Hi Jozef,

Changes applied. Regarding each note:

### Manual review note 1
Answered here: https://github.com/kiegroup/kie-wb-common/pull/3000#issuecomment-553457027

### Manual review note 2
Fixed.

### Manual review note 3
This is Monaco default behaviour, personally it seems quite expected to me, but I've created a JIRA https://issues.jboss.org/browse/DROOLS-4757, then we can explore this subject more there.

### Manual review note 4, note 5
The code completion is enabled only for the Literal Expressions editor (Boxed Expression editor) in this first MVP, not for all input where users can type literal FEEL expressions.

![2019-11-14 09 35 39](https://user-images.githubusercontent.com/1079279/68857920-ab903c00-06c2-11ea-806f-ab9c7248438a.gif)

### Manual review note 6
It will be addressed here https://issues.jboss.org/browse/KOGITO-598

### Manual review note 7
Fixed.

### Manual review note 8
Thanks for reporting. I've created this JIRA to handle it separately https://issues.jboss.org/browse/DROOLS-4758.

### Manual review note 9
Thanks for reporting. I've created this JIRA to handle it separately https://issues.jboss.org/browse/DROOLS-4759.

## Manual review note 10
Thanks for reporting. I've created this JIRA to handle it separately https://issues.jboss.org/browse/DROOLS-4760.

karreiro commented:
@jomarko Here's the [.war for kie-wb-distributions (business-central-webapp)](https://www.dropbox.com/s/5ajitpafknva4as/business-central-webapp-2.war?dl=0).

karreiro commented:
Jenkins, please retest this.

karreiro commented:
(rebasing the branch...)

karreiro commented:
Jenkins, please retest this.

karreiro commented:
(rebasing the branch...)

karreiro commented:
@jomarko Could please take a look?

karreiro commented:
@jomarko Could you please take a look?

karreiro commented:
(rebasing the branch...)

karreiro commented:
@jomarko Thank you for reporting. I've fixed the Firefox issue:

![2019-12-19 16 11 10](https://user-images.githubusercontent.com/1079279/71202189-71462b80-227a-11ea-9dde-2f4582593582.gif)

karreiro commented:
@jomarko I've updated this PR.

Now, when the suggestion widget is opened, and users press "Esc," the suggestion widget closed. Also, when the suggestion widget is closed, and users press "Esc" (again), the edit mode is disabled. See:

![2020-01-02 17 30 48](https://user-images.githubusercontent.com/1079279/71691289-b0e63a80-2d85-11ea-82d4-448a96a5c44e.gif)

karreiro commented:
@manstis I think this ensemble is ready to merge :-)
